src/app/clusters/zone-management-server/zone-util.h

```diff
@@ -0,0 +1,250 @@
+/*
+ *
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/*
+ * Utility Class providing functionality around various checks required for
+ * zones, such as self-intersection, duplicates, etc.
+ *
+ */
+
+#pragma once
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <set>
+#include <vector>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ZoneManagement {
+
+enum class OrientationEnum : uint8_t
+{
+    kCollinear = 0x00,
+    kOnLeft    = 0x01,
+    kOnRight   = 0x02,
+};
+
+class ZoneUtil
+{
+public:
+    using TwoDCartesianVertexStruct = Structs::TwoDCartesianVertexStruct::Type;
+
+    static bool AreTwoDCartVerticesEqual(const TwoDCartesianVertexStruct & v1, const TwoDCartesianVertexStruct & v2)
+    {
+        return v1.x == v2.x && v1.y == v2.y;
+    }
+
+    /**
+     * Algorithm for Checking Self-Intersection of a Zone
```

Comment:
This needs some proof-like thing that this algorithm actually works correctly.

Reply:
Sure. Will provide a proof here. 
Just to highlight the 2 changes that were added to the initial algorithm:
1) Filtering polygons with repeated vertices as a basic prerequisite for _simple_ polygons. This handles the vast majority of degenerate cases.
2) The special case of allowing self-intersection checks on edges of the 3-vertex polygon(with all edges adjacent to each other) which is required because of the optimized logic of checking for self-intersection _only_ among non-adjacent edges.

Reply:
Added Proof of Correctness below the Algorithm outline.

Reply:
Modified the algorithm slightly to have the 3-vertex-only case handled up-front and flagging it as self-intersecting if they are collinear. The algorithm of picking non-adjacent edges, thus, runs on all polygons > 3 vertices.